src/rdb.c

```diff
@@ -1510,9 +1510,13 @@ void rdbRemoveTempFile(pid_t childpid, int from_signal) {
 
     /* Generate temp rdb file name using async-signal safe functions. */
     int pid_len = ll2string(pid, sizeof(pid), childpid);
-    strcpy(tmpfile, "temp-");
-    strncpy(tmpfile+5, pid, pid_len);
-    strcpy(tmpfile+5+pid_len, ".rdb");
+    if (pid_len > 0) {
+        redis_strlcpy(tmpfile, "temp-", sizeof(tmpfile));
+        redis_strlcat(tmpfile, pid, sizeof(tmpfile));
+        redis_strlcat(tmpfile, ".rdb", sizeof(tmpfile));
+    } else {
+        redis_strlcpy(tmpfile,"temp.rdb",sizeof(tmpfile));
```

Comment:
why would pid_len ever be 0?
is the name `temp.rdb` valid in that case?

Reply:
I only meant to cover some theoretical case in which we cannot convert the pid to string. I can also just assert in this case instead.

Reply:
this theoretical case will never really happen, right? in that case i'd rather not have a specific code flow that handles it.
i'd rather either add an assert, or just ignore the case since i think the code would still work ohay even when `pid` is an empty string.

however, i look at the code of ull2string and it doesn't look like it null-terminates the string (not in the failed case, but also not in the successful one).
so doesn't it mean the current code is completely broken?
if we make sure to add a null-term to `pid` at `pid_len` then i think there's no need to handle the special case of 0.

Reply:
@oranagra 
> this theoretical case will never really happen, right? in that case i'd rather not have a specific code flow that handles it.
i'd rather either add an assert, or just ignore the case since i think the code would still work ohay even when pid is an empty string.

I agree that given the fact that pid_t is unsigned int it should never be bigger than 10 digits, so 32 is sufficient enough to hold it. I will remove the error handling case.

> however, i look at the code of ull2string and it doesn't look like it null-terminates the string (not in the failed case, but also not in the successful one).
so doesn't it mean the current code is completely broken?

Well I think the null handling is done [here](https://github.com/redis/redis/blob/unstable/src/util.c#L370) so IMO this code is not broken

Reply:
ok, i missed it last night.
but still it means that in the unsuccessful case (which should be dead code), we don't null-term the `pid` string, and then strlcat will cause a mess.
how about improving `ull2string` and `ll2string` to write a null term in the error cases? i suppose it'll make other uses of them safer too.

Reply:
> ok, i missed it last night. but still it means that in the unsuccessful case (which should be dead code), we don't null-term the `pid` string, and then strlcat will cause a mess. how about improving `ull2string` and `ll2string` to write a null term in the error cases? i suppose it'll make other uses of them safer too.

My opinion is that 0 represent error case in which nothing was written to the buffer and it was left untouched, so I see no point in setting some null termination on the output buffer. it is application responsibility to check for the return value and act accordingly.

Reply:
you're right, but it's a common issue for applications not to check things (that's what this PR is all about, :smile: ), so i think we better attempt to make these two functions safer by writing a null term in the error case as well.

Reply:
@oranagra I will make this change (although I am less keen to the fact that we touch the buffer in case of errors, I agree that this is more safe)